cli/cache/parsed_source.rs

```diff
@@ -0,0 +1,404 @@
+use std::collections::HashMap;
+use std::path::Path;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use deno_ast::MediaType;
+use deno_ast::ModuleSpecifier;
+use deno_ast::ParsedSource;
+use deno_core::error::AnyError;
+use deno_core::parking_lot::Mutex;
+use deno_core::serde_json;
+use deno_graph::CapturingModuleParser;
+use deno_graph::DefaultModuleAnalyzer;
+use deno_graph::ModuleInfo;
+use deno_graph::ModuleParser;
+use deno_graph::ParsedSourceStore;
+use deno_runtime::deno_webstorage::rusqlite::params;
+use deno_runtime::deno_webstorage::rusqlite::Connection;
+
+use super::common::run_sqlite_pragma;
+use super::FastInsecureHasher;
+
+#[derive(Clone, Default)]
+struct ParsedSourceCacheSources(
+  Arc<Mutex<HashMap<ModuleSpecifier, ParsedSource>>>,
+);
+
+/// It's ok that this is racy since in non-LSP situations
+/// this will only ever store one form of a parsed source
+/// and in LSP settings the concurrency will be enforced
+/// at a higher level to ensure this will have the latest
+/// parsed source.
+impl deno_graph::ParsedSourceStore for ParsedSourceCacheSources {
+  fn set_parsed_source(
+    &self,
+    specifier: deno_graph::ModuleSpecifier,
+    parsed_source: ParsedSource,
+  ) -> Option<ParsedSource> {
+    self.0.lock().insert(specifier, parsed_source)
+  }
+
+  fn get_parsed_source(
+    &self,
+    specifier: &deno_graph::ModuleSpecifier,
+  ) -> Option<ParsedSource> {
+    self.0.lock().get(specifier).cloned()
+  }
+}
+
+/// A cache of `ParsedSource`s, which may be used with `deno_graph`
+/// for cached dependency analysis.
+#[derive(Clone)]
+pub struct ParsedSourceCache {
```

Comment:
todo: unit test this

Reply:
This is actually kind of hard to unit test because it's more wiring up different structs. Probably the tests on `ParsedSourceCacheModuleAnalyzer` along with the integration tests are sufficient.